tempesta_fw/http.c

```diff
@@ -197,30 +182,49 @@ tfw_http_prep_302(TfwHttpMsg *hm, TfwStr *cookie)
 	tfw_http_prep_date(__TFW_STR_CH(&rh, 1)->ptr);
 	tfw_http_msg_write(&it, resp, &rh);
 	/*
-	 * HTTP/1.0 can have no host part, so we create relative URI.
+	 * HTTP/1.0 may have no host part, so we create relative URI.
 	 * See RFC 1945 9.3 and RFC 7231 7.1.2.
 	 */
 	if (host.len) {
 		tfw_http_msg_write(&it, resp, &(TfwStr){ .ptr = S_HTTP,
-							 .len = SLEN(S_HTTP)});
+							 .len = SLEN(S_HTTP) });
 		tfw_http_msg_write(&it, resp, &host);
 	}
 	tfw_http_msg_write(&it, resp, &req->uri_path);
 	tfw_http_msg_write(&it, resp, &(TfwStr){ .ptr = S_302_PART_03,
-						 .len = SLEN(S_302_PART_03)});
+						 .len = SLEN(S_302_PART_03) });
 	tfw_http_msg_write(&it, resp, cookie);
-	tfw_http_msg_write(&it, resp, &(TfwStr){ .ptr = S_302_PART_04,
-						 .len = SLEN(S_302_PART_04)});
+	tfw_http_msg_write(&it, resp, crlf);
 
 	return resp;
 }
 
+/*
+ * Perform operations common to sending an error response to a client.
+ * Set current date in the header of an HTTP error response, and set
+ * the "Connection:" header field if it was present in the request.
+ * NOTE: The last chunk of @msg must be CRLF.
+ */
 static int
-tfw_http_send_resp(TfwHttpMsg *hm, const TfwStr *msg, const TfwStr *date)
+tfw_http_send_resp(TfwHttpMsg *hmreq, TfwStr *msg, const TfwStr *date)
 {
+	int conn_flag = hmreq->flags & __TFW_HTTP_CONN_MASK;
+	TfwStr *crlf = __TFW_STR_CH(msg, TFW_STR_CHUNKN(msg) - 1);
 	TfwHttpMsg *resp;
 	TfwMsgIter it;
 
+	if (conn_flag) {
+		unsigned long crlf_len = crlf->len;
+		if (conn_flag == TFW_HTTP_CONN_CLOSE) {
+			crlf->ptr = S_H_CONN_CLOSE;
+			crlf->len = SLEN(S_H_CONN_CLOSE);
+		} else if (conn_flag == TFW_HTTP_CONN_KA) {
```

Comment:
Currently we support only few values for Connection header, while it's much more complex. So other flags will be introduced and it's better to check the flags by ```&```.